### DISCUSSION TITLE
Announcing keycloak-graphql -- A GraphQL API for Keycloak

### DISCUSSION THREAD

**Discussion by blevine** [keycloak/keycloak#34594] [Ideas]
Hello Keycloak aficionados,

I’d like to announce a new open source project I’ve started called: [graphql-keycloak](https://github.com/blevine/keycloak-graphql). This is a GraphQL version of the [Keycloak Admin REST API](https://www.keycloak.org/docs-api/latest/rest-api/index.html). The idea for this project came to me after a few years of building a user management system based on data stored in our Keycloak server. I found it difficult to query this data in a performant way using the REST API and so I thought a GraphQL implementation might solve some of the problems I encountered.

I still have some work to do before I open up to pull requests, but please feel free to check it out and kick the tires. As is mentioned in the README, keycloak-graphql is in its formative stage and should not be used in production at this time. I’m making it public “early” for a couple of reasons:

- I’d like to get some feedback now on how you view the usefulness of this project.
- I’ve integrated a WebSocket server to support GraphQL subscriptions using Vert.x. However, I’m looking for a better way to integrate WebSockets into Keycloak. Any suggestions in this area would be appreciated (see the README).
- As this is my first open source project, I’d like your feedback on the repo in general: layout, README, code structure, build, etc.

Thank you all in advance for your constructive feedback.

**Comment by shawkins** [comment]
Thank you for highlighting this work @blevine. This does seem like a reasonable approach to GraphQL support. I see that there is also a smallrye graphql extension available for quarkus. It might be good to highlight some of the reasoning for using a different stack.

> However, I’m looking for a better way to integrate WebSockets into Keycloak.

The quarkus team would probably have some pointers on this. If it requires another extension, that would unfortunately be a custom build of keycloak as extensions cannot be added via the provider directory.

> I’d like to get some feedback now on how you view the usefulness of this project.

I don't think we've seen much interest in GraphQL directly in Keycloak based upon searching issues and discussions. Some may have tried to layer GraphQL over our rest api, rather that doing this deeper integration. You can keep us posted if you find others interested in your project.

**Reply by blevine** [reply]
Thanks for the feedback.

- The ability to enhance the Keycloak server with Quarkus extensions (whether as part of a Keycloak extension or other mechanism) would really be a boon for developers. I'm sure I'm not the first person to have said that.
- I did look at smallrye graphql at the start. To be honest, I don't remember why I chose graphql-spqr/graphql-java over smallrye.

**Reply by jonkoops** [reply]
I think the main appeal is the possibility of a better quality API. The current REST API is very much built to facilitate the various UIs in Keycloak, and is therefore quite a poor implementation for a generic API to interact with Keycloak. Various aspects such as error handling, pagination, filtering and expanding data structures are all implemented inconsistently across the API.

**Reply by blevine** [reply]
@jonkoops Agreed!

**Reply by shawkins** [reply]
> I'm sure I'm not the first person to have said that.

It's been brought up, but it would need to be addressed at the quarkus level and it would likely be strictly an upstream mechanism.
